microsoft/FluidFramework
# GC: Two-Stage Auto-recovery on Tombstone Load error

## Description

Fixes AB#6234.

During the Grace Period, interpret usage of a Tombstoned object as an added reference, so the unreferenced timestamp will be reset in the next GC (See #18559 for info on Two-Stage Sweep). This will prevent Sweep from deleting it, which is the primary goal.

Note: It's unlikely that this will _permanently_ repair an object that is hitting Tombstone Loaded events.  Two possibilities and outcomes:

1. There is truly no handle to this object, and some app logic is wrongly referencing it.  Until that app code is fixed up, this is liable to repeat periodically (per the Tombstone Timeout).  But it gives app devs a chance to observe the bug while protecting end users from permanent harm (assuming they hit this during the Grace Period)
2. There is a proper reference to the object, but the GC data was miscalculated or corrupted.  Unless the object with the reference changes and has its GC data recomputed properly, this will also repeat periodically.  The plan is to make a subsequent PR that will also trigger Full GC on the next GC run after the TombstoneLoaded op.

### Implementation Plan

The Tombstone Load error will occur first on one particular client - likely an Interactive client - but all clients, especially the Summarizer, need to reset the unreferenced tracking.  So we need a way to notify all clients when this happens.

Conveniently, we recently added a new GC op type (See #18632) which is extensible with sub-types, so we can add a new GC op subtype for when a Tombstone is loaded.  When each client processes this op it can simply call `addedOutboundReference`. This will be a functional no-op on Interactive clients (besides logging) but for Summarizer clients it will cause that object and any objects it references to have their unreferenced timestamps reset.

As mentioned above, it's possible (even likely) that after Tombstone Timeout this process will repeat itself.

### Test Plan

The new behavior can be tested by a single end-to-end test following this script:

* Create two DataStores A and B, with A referencing B and Default referencing A.
* Unreference A and let A and B become tombstones
* Request A, triggering the TombstoneLoaded op
* Once the op has been processed, check for the `GC_Tombstone_DataStore_Revived` logging event
* Observe that A is still tombstoned in the existing interactive client (since that state doesn't update)
* Summarize now that the op roundtripped, load a new container from that summary. (Auto-Recovery would be in effect)
* Confirm that neither A nor B are tombstones anymore.  (but remember they are unreferenced)
* Wait the Tombstone Timeout again, and confirm that A and B are tombstones again.

## Reviewer Guidance

Please review the test changes carefully - Thanks!

`//*` comments are points to discuss while reviewing the code.

To Do:

- [x] Write up test plan
- [x] Address `//*` Todo/question comments throughout
- [x] Update tests that are made more difficult via auto-recovery to just disable it (gives coverage of feature flag too)

Punted for now:

- Add test coverage of Summarizer not submitting the op
  - It complicates the feature, and Navin is thinking holistically about GC op emitted from Summarizer (this was the motivation to avoid it)
- Add test coverage of direct `GC_Tombstone_SubDataStore_Requested` (DDS) case
  - Primary 1P consumer doesn't use this kind of access pattern, and it's clear in the code that this case is ok.  Will open a follow-up task to revisit.
